src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/PrisonerCreationService.kt

```diff
@@ -95,28 +95,28 @@ class PrisonerCreationService(
     return PrisonerIdentifier().withId(currentSequence.prisonerIdentifier)
   }
 
-  private fun gender(code: String): Result<Gender> =
-    genderRepository.findByIdOrNull(ReferenceCode.Pk(Gender.SEX, code))?.let { success(it) } ?: failure(
-      EntityNotFoundException.withMessage("Gender $code not found")
-    )
+  private fun gender(code: String): Result<Gender> = genderRepository.getMandatoryReference(code, Gender.SEX)
 
-  private fun ethnicity(code: String?): Result<Ethnicity>? =
-    code?.takeIf { it.isNotBlank() }?.let {
-      ethnicityRepository.findByIdOrNull(ReferenceCode.Pk(Ethnicity.ETHNICITY, code))?.let { success(it) }
-        ?: failure(EntityNotFoundException.withMessage("Ethnicity $code not found"))
-    }
+  private fun title(code: String?): Result<Title?> = titleRepository.getOptionalReference(code, Title.TITLE)
 
-  private fun title(code: String?): Result<Title>? =
-    code?.takeIf { it.isNotBlank() }?.let {
-      titleRepository.findByIdOrNull(ReferenceCode.Pk(Title.TITLE, code))?.let { success(it) }
-        ?: failure(EntityNotFoundException.withMessage("Title $code not found"))
-    }
+  private fun suffix(code: String?): Result<Suffix?> = suffixRepository.getOptionalReference(code, Suffix.SUFFIX)
+
+  private fun ethnicity(code: String?): Result<Ethnicity?> = ethnicityRepository.getOptionalReference(code, Ethnicity.ETHNICITY)
+
+  private inline fun <reified T: ReferenceCode> ReferenceCodeRepository<T>.getMandatoryReference(code: String, pk: String): Result<T> =
+    findByIdOrNull(ReferenceCode.Pk(pk, code))
+      ?.let { success(it) }
+      ?: failure(EntityNotFoundException.withMessage("${T::class.java} $code not found"))
+
+  private inline fun <reified T: ReferenceCode> ReferenceCodeRepository<T>.getOptionalReference(code: String?, pk: String): Result<T?> =
+    code?.takeIf { it.isNotBlank() }
+      ?.let {
+        findByIdOrNull(ReferenceCode.Pk(pk, code))
+          ?.let { success(it) }
+          ?: failure(EntityNotFoundException.withMessage("${T::class.java} $code not found"))
+      }
+      ?: success(null)
```

Comment:
Just wanted to see what it looks like the other way around. I think I prefer this as it is explicitly says null is a success, and the extra question mark on the caller kept making me think twice. But it's 6 and two 3s really.